File: server/src/main/scala/com/mitrakov/self/serverpush/FirebaseHelper.scala
package com.mitrakov.self.serverpush

import com.google.auth.oauth2.GoogleCredentials
import com.google.firebase.{FirebaseApp, FirebaseOptions}
import com.google.firebase.messaging.{ApnsConfig, Aps, FirebaseMessaging, Message, Notification}
import com.typesafe.scalalogging.StrictLogging
import java.io.FileInputStream

class FirebaseHelper(secretConfig: String) extends StrictLogging {
  type MsgId = String

  val projectId = "tommypush-405b7"
  val credentials: GoogleCredentials = GoogleCredentials.fromStream(new FileInputStream(secretConfig))
  val options: FirebaseOptions = FirebaseOptions.builder().setCredentials(credentials).setProjectId(projectId).build()
  val app: FirebaseApp = FirebaseApp.initializeApp(options)

  def sendMessage(recipientFcmToken: String, title: String, body: String): MsgId = {
    logger.info("Sending message {}: {}", title, body)
    val notification = Notification.builder().setTitle(title).setBody(body).build()
    val apnsConfig = ApnsConfig.builder().setAps(Aps.builder().setSound("default").build()).build()
    val msg = Message.builder().setNotification(notification).setApnsConfig(apnsConfig).setToken(recipientFcmToken).build()

    FirebaseMessaging.getInstance(app).send(msg)
  }
}

File: server/src/main/java/com/mitrakoff/self/tommypush/FirebaseHelper.java
package com.mitrakoff.self.tommypush;

import com.google.auth.oauth2.GoogleCredentials;
import com.google.firebase.*;
import com.google.firebase.messaging.*;
import java.io.IOException;
import java.nio.file.*;
import java.time.LocalDateTime;

public class FirebaseHelper implements AutoCloseable {
    public final FirebaseApp app;

    public FirebaseHelper(String secretConfig) {
        try {
            final GoogleCredentials credentials = GoogleCredentials.fromStream(Files.newInputStream(Paths.get(secretConfig)));
            final FirebaseOptions options = FirebaseOptions.builder().setCredentials(credentials).build();
            app = FirebaseApp.initializeApp(options);
        } catch (IOException e) {throw new RuntimeException(e);}
    }

    /** @return messageID from Firebase Cloud Messaging */
    public String sendMessage(String recipientFcmToken, String title, String body) {
        System.out.printf("%s: Sending message %s: %s\n", LocalDateTime.now(), title, body);
        final Notification notification = Notification.builder().setTitle(title).setBody(body).build();
        final ApnsConfig apnsConfig = ApnsConfig.builder().setAps(Aps.builder().setSound("default").build()).build();
        final Message message = Message.builder().setNotification(notification).setApnsConfig(apnsConfig).setToken(recipientFcmToken).build();

        try {
            return FirebaseMessaging.getInstance(app).send(message);
        } catch (FirebaseMessagingException e) {e.printStackTrace(); return "";}
    }

    @Override
    public void close() { app.delete(); }
}
